torchao/prototype/autoround/core.py

```diff
@@ -0,0 +1,391 @@
+            self.count += 1
+            self.values.append(input)
+        return self
+
+    def pad_to_length(self, length):
+        if self.count > length:
+            return self
+        self.add_tensors([self.values[-1]] * (length - self.count))
+        return self
+
+    @classmethod
+    def flat_to_grouped(cls, flat):
+        # size of biggest MultiTensor
+        multi_tensor_size = max(
+            [x.count if isinstance(x, MultiTensor) else 1 for x in flat]
+        )
+        # convert [A, MultiTensor(b1,b2,b3), MultiTensor(c1,c2,c3)] => [[A,b1,c1], [A,b2,c2] [A,b3,c3]]
+        grouped = list(
+            zip(
+                *[
+                    (
+                        x.pad_to_length(multi_tensor_size).values
+                        if isinstance(x, MultiTensor)
+                        else [x] * multi_tensor_size
+                    )
+                    for x in flat
+                ]
+            )
+        )
+        return grouped
+
+    @classmethod
+    def grouped_to_flat(cls, grouped):
+        # convert [[A,b1,c1], [A,b2,c2] [A,b3,c3]] => [A, MultiTensor(b1,b2,b3), MultiTensor(c1,c2,c3)]
+        # where A is nontensor, b's,c's are tensors
+        # convert [[A,b1,c1], [A,b2,c2] [A,b3,c3]] => [(A,A,A), (b1,b2,b3), (c1,c2,c3)]
+        flat_tups = list(zip(*grouped))
+        # convert [(A,A,A), (b1,b2,b3), (c1,c2,c3)] => [A, MultiTensor(b1,b2,b3), MultiTensor(c1,c2,c3)]
+        flattened = [
+            cls(tup).cpu() if isinstance(tup[0], torch.Tensor) else tup[0]
+            for tup in flat_tups
+        ]
+        # need to check that getting rid of all but one from each nonTensor tuple is OK
+        non_tensors_equal = min(
+            [True]
+            + [
+                min(
+                    [True]
+                    + [  # handle situation where tuples have size 0
+                        tup[0] == x for x in tup  # check all elements match
+                    ]
+                )
+                for tup in flat_tups
+                if not isinstance(tup[0], torch.Tensor)  # look at tuples of nonTensors
+            ]
+        )
+        return flattened, non_tensors_equal
+
+    @classmethod
+    def __torch_function__(cls, func, types, args=(), kwargs=None, skip_gptq=False):
+        kwargs = {} if kwargs is None else kwargs
+        # combine args and kwargs and remove lists and tuples
+        flat_args, spec = tree_flatten((args, kwargs))
+        # convert [A, MultiTensor(b1,b2,b3), MultiTensor(c1,c2,c3)] => [[A,b1,c1], [A,b2,c2] [A,b3,c3]]
+        grouped_args = cls.flat_to_grouped(flat_args)
+        # run function for each of the multitensors and return a multitensor
+        outputs = []
+        with torch._C.DisableTorchFunctionSubclass():
+            # Note: for the decoder, apply auto-round to it
+            if func is torch.ops.transformers_ops.general_decoder:
```

Comment:
Hi @jerryzh168, Hi, I am trying to follow your [suggestion](https://github.com/pytorch/ao/pull/581#issuecomment-2264104863) to use `_implements`  for `MulTensor`, but it seems that it only supports registering one or several ops at a time.  However in the `auto-round` scenario, except for `general_decoder`, the other implementations are basically the same, like below:

```python

@_implements([torch.ops.transformers_ops.general_decoder])
def _(...):
    ...

# For the other ops 
@_implements([*])
def _(func, types, args, kwargs):
    ...
    for i, inp in enumerate(grouped_args):
        cur_args, cur_kwargs = tree_unflatten(inp, spec)
        out = func(*cur_args, **cur_kwargs)
        outputs.append(out.cpu() if isinstance(out, torch.Tensor) else out)
```
Can we extend `_implements` to support wildcards?

Reply:
ah I see, yeah I can add that